### PR TITLE
fixes get ETCD version from kubernetes constants #11290

### PR DESCRIPTION
**Pull request by Srikrishnabh**
fix for #11290

### CODE REVIEW

File: pkg/minikube/constants/constants_kubeadm_images.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package constants
+
+var (
+	KubeadmImages = map[string]map[string]string{
+		"v1.22": {
```
Comment (afbjorklund) [inline pkg/minikube/constants/constants_kubeadm_images.go:21]: These versions look wrong (1.21.3, etc)

**Review by afbjorklund — CHANGES_REQUESTED** [review]

**Review by medyagh — COMMENTED** [review]
This pr looks good.

**Comment by medyagh** [comment]
@Srikrishnabh would you plz rebase

**Comment by Srikrishnabh** [comment]
> @Srikrishnabh would you plz rebase

@medyagh rebased.

File: hack/update/kubeadm_constants/update_kubeadm_constants.go [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+func executeCommand(command string, args ...string) (string, error) {
+	output, err := exec.Command(command, args...).Output()
+	if err != nil {
+		return "", err
+	}
+	return string(output), nil
+}
+
+// getK8sVersion returns Kubernetes versions.
+func getK8sVersions(ctx context.Context, owner, repo string) (stable, latest, latestMM, latestP0 string, err error) {
```
Comment (spowelljr) [inline hack/update/kubeadm_constants/update_kubeadm_constants.go:193]: `latestMM` & `latestP0` aren't used anywhere, couldn't we just replace this whole function with a direct call to `update.GHReleases`?

File: pkg/minikube/bootstrapper/images/kubeadm_test.go [hunk 3]
```diff
@@ -46,17 +46,17 @@ func TestKubeadmImages(t *testing.T) {
...
-			"mirror.k8s.io/kube-scheduler:v1.16.1",
-			"mirror.k8s.io/kube-controller-manager:v1.16.1",
-			"mirror.k8s.io/kube-apiserver:v1.16.1",
-			"mirror.k8s.io/coredns:1.6.2",
-			"mirror.k8s.io/etcd:3.3.15-0",
-			"mirror.k8s.io/pause:3.1",
-			"mirror.k8s.io/k8s-minikube/storage-provisioner:" + version.GetStorageProvisionerVersion(),
-			"mirror.k8s.io/kubernetesui/dashboard:v2.3.1",
-			"mirror.k8s.io/kubernetesui/metrics-scraper:v1.0.7",
+		{"v1.16.1", "k8s.gcr.io", false, []string{
```
[the remark this answers is not in the document]
Comment (Srikrishnabh) [reply pkg/minikube/bootstrapper/images/kubeadm_test.go]: This was due to constants having the k8s.gcr.io as fixed repo. I have fixed this. Thanks.

File: pkg/minikube/bootstrapper/images/images.go [hunk 4]
```diff
@@ -73,30 +71,20 @@ func coreDNS(v semver.Version, mirror string) string {
...
-		cv = "1.2.2"
+
+	majorMinorVersion := fmt.Sprintf("v%d.%d", v.Major, v.Minor)
+	imageName := path.Join(kubernetesRepo(mirror), in)
+	if cVersion, ok := constants.KubeadmImages[majorMinorVersion][imageName]; ok {
+		cv = cVersion
 	}
-	return path.Join(kubernetesRepo(mirror), in+":"+cv)
+
+	return fmt.Sprintf("%s:%s", imageName, cv)
```
[the remark this answers is not in the document]
Comment (Srikrishnabh) [reply pkg/minikube/bootstrapper/images/images.go]: fixed it.

**Review by medyagh — CHANGES_REQUESTED** [review]
I just tried to run this on my Mac M1 Laptop Arm64 but I got this error

I believe this needs to download arm64 binary instead
```
1:03:55 medya/workspace/minikube
master_srikrishnabh ⌅1
$ make update-kubeadm-constants 
(cd hack/update/kubeadm_constants && \
	 go run update_kubeadm_constants.go)
E1123 11:04:11.685666   47743 update_kubeadm_constants.go:122] failed to execute kubeadm command ./kubeadm-linux-amd64-v1.22.4
F1123 11:04:11.689208   47743 update_kubeadm_constants.go:84] fork/exec ./kubeadm-linux-amd64-v1.22.4: exec format error
goroutine 1 [running]:
k8s.io/klog/v2.stacks(0x1)
	/Users/medya/go/pkg/mod/k8s.io/klog/v2@v2.30.0/klog.go:1038 +0xac
k8s.io/klog/v2.(*loggingT).output(0x101978340, 0x3, 0x0, 0x1400027a000, 0x0, {0x101686f83, 0x1b}, 0x54, 0x0)
	/Users/medya/go/pkg/mod/k8s.io/klog/v2@v2.30.0/klog.go:987 +0x628
k8s.io/klog/v2.(*loggingT).println(0x101978340, 0x3, 0x0, {0x0, 0x0}, {0x1400035dfb0, 0x1, 0x1})
	/Users/medya/go/pkg/mod/k8s.io/klog/v2@v2.30.0/klog.go:713 +0x14c
k8s.io/klog/v2.Fatalln(...)
	/Users/medya/go/pkg/mod/k8s.io/klog/v2@v2.30.0/klog.go:1525
main.main()
	/Users/medya/workspace/minikube/hack/update/kubeadm_constants/update_kubeadm_constants.go:84 +0x544

goroutine 19 [chan receive]:
k8s.io/klog/v2.(*loggingT).flushDaemon(0x101978340)
	/Users/medya/go/pkg/mod/k8s.io/klog/v2@v2.30.0/klog.go:1181 +0x70
created by k8s.io/klog/v2.init.0
	/Users/medya/go/pkg/mod/k8s.io/klog/v2@v2.30.0/klog.go:420 +0x150

goroutine 8 [IO wait]:
internal/poll.runtime_pollWait(0x101d9e4f8, 0x72)
	/usr/local/go/src/runtime/netpoll.go:229 +0xa4
internal/poll.(*pollDesc).wait(0x1400021a698, 0x72, 0x0)
	/usr/local/go/src/internal/poll/fd_poll_runtime.go:84 +0x38
internal/poll.(*pollDesc).waitRead(...)
	/usr/local/go/src/internal/poll/fd_poll_runtime.go:89
internal/poll.(*FD).Read(0x1400021a680, {0x140001be000, 0x2053, 0x2053})
	/usr/local/go/src/internal/poll/fd_unix.go:167 +0x1dc
net.(*netFD).Read(0x1400021a680, {0x140001be000, 0x2053, 0x2053})
	/usr/local/go/src/net/fd_posix.go:56 +0x44
net.(*conn).Read(0x14000010010, {0x140001be000, 0x2053, 0x2053})
	/usr/local/go/src/net/net.go:183 +0x4c
crypto/tls.(*atLeastReader).Read(0x140001860c0, {0x140001be000, 0x2053, 0x2053})
	/usr/local/go/src/crypto/tls/conn.go:777 +0x58
bytes.(*Buffer).ReadFrom(0x14000300278, {0x1015b2340, 0x140001860c0})
	/usr/local/go/src/bytes/buffer.go:204 +0xa4
crypto/tls.(*Conn).readFromUntil(0x14000300000, {0x128b30008, 0x14000010010}, 0x5)
	/usr/local/go/src/crypto/tls/conn.go:799 +0xe0
crypto/tls.(*Conn).readRecordOrCCS(0x14000300000, 0x0)
	/usr/local/go/src/crypto/tls/conn.go:606 +0xf4
crypto/tls.(*Conn).readRecord(...)
	/usr/local/go/src/crypto/tls/conn.go:574
crypto/tls.(*Conn).Read(0x14000300000, {0x14000478000, 0x1000, 0x1000})
	/usr/local/go/src/crypto/tls/conn.go:1277 +0x164
bufio.(*Reader).Read(0x14000470060, {0x14000423538, 0x9, 0x9})
	/usr/local/go/src/bufio/bufio.go:227 +0x20c
io.ReadAtLeast({0x1015b21e0, 0x14000470060}, {0x14000423538, 0x9, 0x9}, 0x9)
	/usr/local/go/src/io/io.go:328 +0xa0
io.ReadFull(...)
	/usr/local/go/src/io/io.go:347
net/http.http2readFrameHeader({0x14000423538, 0x9, 0x9}, {0x1015b21e0, 0x14000470060})
	/usr/local/go/src/net/http/h2_bundle.go:1553 +0x5c
net/http.(*http2Framer).ReadFrame(0x14000423500)
	/usr/local/go/src/net/http/h2_bundle.go:1811 +0x90
net/http.(*http2clientConnReadLoop).run(0x1400005ffa8)
	/usr/local/go/src/net/http/h2_bundle.go:8437 +0x13c
net/http.(*http2ClientConn).readLoop(0x1400007e180)
	/usr/local/go/src/net/http/h2_bundle.go:8359 +0x60
created by net/http.(*http2Transport).newClientConn
	/usr/local/go/src/net/http/h2_bundle.go:7311 +0xb48

goroutine 26 [IO wait]:
internal/poll.runtime_pollWait(0x101d9e410, 0x72)
	/usr/local/go/src/runtime/netpoll.go:229 +0xa4
internal/poll.(*pollDesc).wait(0x140004a5398, 0x72, 0x0)
	/usr/local/go/src/internal/poll/fd_poll_runtime.go:84 +0x38
internal/poll.(*pollDesc).waitRead(...)
	/usr/local/go/src/internal/poll/fd_poll_runtime.go:89
internal/poll.(*FD).Read(0x140004a5380, {0x14000594000, 0xe5a2, 0xe5a2})
	/usr/local/go/src/internal/poll/fd_unix.go:167 +0x1dc
net.(*netFD).Read(0x140004a5380, {0x14000594000, 0xe5a2, 0xe5a2})
	/usr/local/go/src/net/fd_posix.go:56 +0x44
net.(*conn).Read(0x14000132720, {0x14000594000, 0xe5a2, 0xe5a2})
	/usr/local/go/src/net/net.go:183 +0x4c
crypto/tls.(*atLeastReader).Read(0x1400037af30, {0x14000594000, 0xe5a2, 0xe5a2})
	/usr/local/go/src/crypto/tls/conn.go:777 +0x58
bytes.(*Buffer).ReadFrom(0x1400028c5f8, {0x1015b2340, 0x1400037af30})
	/usr/local/go/src/bytes/buffer.go:204 +0xa4
crypto/tls.(*Conn).readFromUntil(0x1400028c380, {0x128b30008, 0x14000132720}, 0x5)
	/usr/local/go/src/crypto/tls/conn.go:799 +0xe0
crypto/tls.(*Conn).readRecordOrCCS(0x1400028c380, 0x0)
	/usr/local/go/src/crypto/tls/conn.go:606 +0xf4
crypto/tls.(*Conn).readRecord(...)
	/usr/local/go/src/crypto/tls/conn.go:574
crypto/tls.(*Conn).Read(0x1400028c380, {0x140002c0000, 0x1000, 0x1000})
	/usr/local/go/src/crypto/tls/conn.go:1277 +0x164
bufio.(*Reader).Read(0x140002bca20, {0x14000206498, 0x9, 0x9})
	/usr/local/go/src/bufio/bufio.go:227 +0x20c
io.ReadAtLeast({0x1015b21e0, 0x140002bca20}, {0x14000206498, 0x9, 0x9}, 0x9)
	/usr/local/go/src/io/io.go:328 +0xa0
io.ReadFull(...)
	/usr/local/go/src/io/io.go:347
net/http.http2readFrameHeader({0x14000206498, 0x9, 0x9}, {0x1015b21e0, 0x140002bca20})
	/usr/local/go/src/net/http/h2_bundle.go:1553 +0x5c
net/http.(*http2Framer).ReadFrame(0x14000206460)
	/usr/local/go/src/net/http/h2_bundle.go:1811 +0x90
net/http.(*http2clientConnReadLoop).run(0x140004b5fa8)
	/usr/local/go/src/net/http/h2_bundle.go:8437 +0x13c
net/http.(*http2ClientConn).readLoop(0x14000260a80)
	/usr/local/go/src/net/http/h2_bundle.go:8359 +0x60
created by net/http.(*http2Transport).newClientConn
	/usr/local/go/src/net/http/h2_bundle.go:7311 +0xb48
exit status 255
make: *** [update-kubeadm-constants] Error 1
11:04:11 medya/workspace/minikube
master_srikrishnabh ⌅2
$ 

```

**Comment by Srikrishnabh** [comment]
> I just tried to run this on my Mac M1 Laptop Arm64 but I got this error
> 
> I believe this needs to download arm64 binary instead
> 
> ```
> 1:03:55 medya/workspace/minikube
> master_srikrishnabh ⌅1
> $ make update-kubeadm-constants 
> (cd hack/update/kubeadm_constants && \
> 	 go run update_kubeadm_constants.go)
> E1123 11:04:11.685666   47743 update_kubeadm_constants.go:122] failed to execute kubeadm command ./kubeadm-linux-amd64-v1.22.4
> F1123 11:04:11.689208   47743 update_kubeadm_constants.go:84] fork/exec ./kubeadm-linux-amd64-v1.22.4: exec format error
> goroutine 1 [running]:
> k8s.io/klog/v2.stacks(0x1)
> 	/Users/medya/go/pkg/mod/k8s.io/klog/v2@v2.30.0/klog.go:1038 +0xac
> k8s.io/klog/v2.(*loggingT).output(0x101978340, 0x3, 0x0, 0x1400027a000, 0x0, {0x101686f83, 0x1b}, 0x54, 0x0)
> 	/Users/medya/go/pkg/mod/k8s.io/klog/v2@v2.30.0/klog.go:987 +0x628
> k8s.io/klog/v2.(*loggingT).println(0x101978340, 0x3, 0x0, {0x0, 0x0}, {0x1400035dfb0, 0x1, 0x1})
> 	/Users/medya/go/pkg/mod/k8s.io/klog/v2@v2.30.0/klog.go:713 +0x14c
> k8s.io/klog/v2.Fatalln(...)
> 	/Users/medya/go/pkg/mod/k8s.io/klog/v2@v2.30.0/klog.go:1525
> main.main()
> 	/Users/medya/workspace/minikube/hack/update/kubeadm_constants/update_kubeadm_constants.go:84 +0x544
> 
> goroutine 19 [chan receive]:
> k8s.io/klog/v2.(*loggingT).flushDaemon(0x101978340)
> 	/Users/medya/go/pkg/mod/k8s.io/klog/v2@v2.30.0/klog.go:1181 +0x70
> created by k8s.io/klog/v2.init.0
> 	/Users/medya/go/pkg/mod/k8s.io/klog/v2@v2.30.0/klog.go:420 +0x150
> 
> goroutine 8 [IO wait]:
> internal/poll.runtime_pollWait(0x101d9e4f8, 0x72)
> 	/usr/local/go/src/runtime/netpoll.go:229 +0xa4
> internal/poll.(*pollDesc).wait(0x1400021a698, 0x72, 0x0)
> 	/usr/local/go/src/internal/poll/fd_poll_runtime.go:84 +0x38
> internal/poll.(*pollDesc).waitRead(...)
> 	/usr/local/go/src/internal/poll/fd_poll_runtime.go:89
> internal/poll.(*FD).Read(0x1400021a680, {0x140001be000, 0x2053, 0x2053})
> 	/usr/local/go/src/internal/poll/fd_unix.go:167 +0x1dc
> net.(*netFD).Read(0x1400021a680, {0x140001be000, 0x2053, 0x2053})
> 	/usr/local/go/src/net/fd_posix.go:56 +0x44
> net.(*conn).Read(0x14000010010, {0x140001be000, 0x2053, 0x2053})
> 	/usr/local/go/src/net/net.go:183 +0x4c
> crypto/tls.(*atLeastReader).Read(0x140001860c0, {0x140001be000, 0x2053, 0x2053})
> 	/usr/local/go/src/crypto/tls/conn.go:777 +0x58
> bytes.(*Buffer).ReadFrom(0x14000300278, {0x1015b2340, 0x140001860c0})
> 	/usr/local/go/src/bytes/buffer.go:204 +0xa4
> crypto/tls.(*Conn).readFromUntil(0x14000300000, {0x128b30008, 0x14000010010}, 0x5)
> 	/usr/local/go/src/crypto/tls/conn.go:799 +0xe0
> crypto/tls.(*Conn).readRecordOrCCS(0x14000300000, 0x0)
> 	/usr/local/go/src/crypto/tls/conn.go:606 +0xf4
> crypto/tls.(*Conn).readRecord(...)
> 	/usr/local/go/src/crypto/tls/conn.go:574
> crypto/tls.(*Conn).Read(0x14000300000, {0x14000478000, 0x1000, 0x1000})
> 	/usr/local/go/src/crypto/tls/conn.go:1277 +0x164
> bufio.(*Reader).Read(0x14000470060, {0x14000423538, 0x9, 0x9})
> 	/usr/local/go/src/bufio/bufio.go:227 +0x20c
> io.ReadAtLeast({0x1015b21e0, 0x14000470060}, {0x14000423538, 0x9, 0x9}, 0x9)
> 	/usr/local/go/src/io/io.go:328 +0xa0
> io.ReadFull(...)
> 	/usr/local/go/src/io/io.go:347
> net/http.http2readFrameHeader({0x14000423538, 0x9, 0x9}, {0x1015b21e0, 0x14000470060})
> 	/usr/local/go/src/net/http/h2_bundle.go:1553 +0x5c
> net/http.(*http2Framer).ReadFrame(0x14000423500)
> 	/usr/local/go/src/net/http/h2_bundle.go:1811 +0x90
> net/http.(*http2clientConnReadLoop).run(0x1400005ffa8)
> 	/usr/local/go/src/net/http/h2_bundle.go:8437 +0x13c
> net/http.(*http2ClientConn).readLoop(0x1400007e180)
> 	/usr/local/go/src/net/http/h2_bundle.go:8359 +0x60
> created by net/http.(*http2Transport).newClientConn
> 	/usr/local/go/src/net/http/h2_bundle.go:7311 +0xb48
> 
> goroutine 26 [IO wait]:
> internal/poll.runtime_pollWait(0x101d9e410, 0x72)
> 	/usr/local/go/src/runtime/netpoll.go:229 +0xa4
> internal/poll.(*pollDesc).wait(0x140004a5398, 0x72, 0x0)
> 	/usr/local/go/src/internal/poll/fd_poll_runtime.go:84 +0x38
> internal/poll.(*pollDesc).waitRead(...)
> 	/usr/local/go/src/internal/poll/fd_poll_runtime.go:89
> internal/poll.(*FD).Read(0x140004a5380, {0x14000594000, 0xe5a2, 0xe5a2})
> 	/usr/local/go/src/internal/poll/fd_unix.go:167 +0x1dc
> net.(*netFD).Read(0x140004a5380, {0x14000594000, 0xe5a2, 0xe5a2})
> 	/usr/local/go/src/net/fd_posix.go:56 +0x44
> net.(*conn).Read(0x14000132720, {0x14000594000, 0xe5a2, 0xe5a2})
> 	/usr/local/go/src/net/net.go:183 +0x4c
> crypto/tls.(*atLeastReader).Read(0x1400037af30, {0x14000594000, 0xe5a2, 0xe5a2})
> 	/usr/local/go/src/crypto/tls/conn.go:777 +0x58
> bytes.(*Buffer).ReadFrom(0x1400028c5f8, {0x1015b2340, 0x1400037af30})
> 	/usr/local/go/src/bytes/buffer.go:204 +0xa4
> crypto/tls.(*Conn).readFromUntil(0x1400028c380, {0x128b30008, 0x14000132720}, 0x5)
> 	/usr/local/go/src/crypto/tls/conn.go:799 +0xe0
> crypto/tls.(*Conn).readRecordOrCCS(0x1400028c380, 0x0)
> 	/usr/local/go/src/crypto/tls/conn.go:606 +0xf4
> crypto/tls.(*Conn).readRecord(...)
> 	/usr/local/go/src/crypto/tls/conn.go:574
> crypto/tls.(*Conn).Read(0x1400028c380, {0x140002c0000, 0x1000, 0x1000})
> 	/usr/local/go/src/crypto/tls/conn.go:1277 +0x164
> bufio.(*Reader).Read(0x140002bca20, {0x14000206498, 0x9, 0x9})
> 	/usr/local/go/src/bufio/bufio.go:227 +0x20c
> io.ReadAtLeast({0x1015b21e0, 0x140002bca20}, {0x14000206498, 0x9, 0x9}, 0x9)
> 	/usr/local/go/src/io/io.go:328 +0xa0
> io.ReadFull(...)
> 	/usr/local/go/src/io/io.go:347
> net/http.http2readFrameHeader({0x14000206498, 0x9, 0x9}, {0x1015b21e0, 0x140002bca20})
> 	/usr/local/go/src/net/http/h2_bundle.go:1553 +0x5c
> net/http.(*http2Framer).ReadFrame(0x14000206460)
> 	/usr/local/go/src/net/http/h2_bundle.go:1811 +0x90
> net/http.(*http2clientConnReadLoop).run(0x140004b5fa8)
> 	/usr/local/go/src/net/http/h2_bundle.go:8437 +0x13c
> net/http.(*http2ClientConn).readLoop(0x14000260a80)
> 	/usr/local/go/src/net/http/h2_bundle.go:8359 +0x60
> created by net/http.(*http2Transport).newClientConn
> 	/usr/local/go/src/net/http/h2_bundle.go:7311 +0xb48
> exit status 255
> make: *** [update-kubeadm-constants] Error 1
> 11:04:11 medya/workspace/minikube
> master_srikrishnabh ⌅2
> $ 
> ```

@medyagh I'm not able to get the kubeadm for mac 1 (darwin-arm64) version. is there a link where I can download it.?

**Comment by medyagh** [comment]
you are right @Srikrishnabh sorry I will try to test this PR on Linux

**Review by medyagh — APPROVED** [review]
this looks good to me thank you @Srikrishnabh

**Comment by Srikrishnabh** [comment]
> this looks good to me thank you @Srikrishnabh

Thank you @medyagh

**Pull request merged**